Add notice asking user for reviews/feedback

fjorgemota opened the pull request:
### Changes proposed in this Pull Request

* Add option `sensei_installed_at` to save when Sensei was installed/updated for the first time;
* Add condition `installed_since` to show notice only after a given time was passed; (I accept feedback regarding this name)
* Add "interactive" dismissible notice asking the user for reviews if some time passed;
* Add way to run extra "tasks" on the frontend like show/hide a notice when clicking on a button/link in a notice;

### Testing instructions

1. Install the plugin based on this branch of the code;
2. Verify that you can visit the Sensei Home page without any issues;
3. Change the URL on the class `Sensei_Home_Remote_Data_API` to point to a WP with this plugin correctly installed 12-gh-Automattic/senseilms-com-plugins 
4. Alter the option `sensei_installed_at` to have a unix timestamp from more than 10 days ago;
5. Verify if the notice asking for reviews appears;
6. Verify that, if you click "Yes", the notice ask the user for a review;
7. Verify that, if you click "No", the notice will ask the user for feedback;
8. Verify that you can dismiss the notice;
9. Verify if the notice, after dismissed, still appears to other site admins;
10. Verify that the notice won't show for users that are not super admins;

### Screenshots

After the user installed Sensei for some time:

![Image showing notice asking the user if it likes Sensei](https://user-images.githubusercontent.com/529864/207099538-1ff118c5-3a6f-475f-acbc-2ec632d6c306.png)

Image showing notice after user clicked on "Yes" option:

![Image showing notice after user clicked on "Yes" option](https://user-images.githubusercontent.com/529864/207099791-5d083016-4f3c-4a38-8bc9-65268b1efd93.png)


Image showing notice after user clicked on "No" option:

![Image showing notice after user clicked on "No" option](https://user-images.githubusercontent.com/529864/207099645-bc1e2bfa-0bc2-44d1-8eda-f1b832eee084.png)

fjorgemota committed:
Add option saving the time when Sensei was installed/updated

fjorgemota committed:
Add support for condition to show notice some time after Sensei was installed

fjorgemota committed:
Add notice asking the user for review

fjorgemota committed:
Add unit tests for the installed_since condition on notices

fjorgemota committed:
Structure added tests for installed_since condition

fjorgemota committed:
Add unit tests for the add_review_notice method

fjorgemota committed:
Fix typo and use $notice_id on the assertArrayHasKey call

fjorgemota committed:
grant_super_admin to check if test pass correctly on Github Actions

fjorgemota committed:
Remove sensei_installed_at option when uninstalling Sensei

fjorgemota committed:
Change messaging and use info links instead of linking directly in the notice

fjorgemota committed:
Set "Yes" button as secondary in the Review notice

fjorgemota committed:
Adjust padding of the notices on Sensei Home

fjorgemota committed:
Center icon on Sensei Notices on Sensei Home

fjorgemota committed:
Various adjustments for the styles of notices on Sensei Home

fjorgemota committed:
Add "success" level to notice adding a green border

fjorgemota committed:
Fix tests to match new messaging on the review notice

fjorgemota committed:
Use add_option instead of update_option so we add sensei_installed_at just once

alexsanford commented on assets/home/notices.scss:
```diff
@@ -28,6 +33,19 @@
 			}
+
+			.button-secondary {
+				color: #155E65;
```
It seems a little unexpected to me that the text in these buttons is different from the border. Was this a deliberate design choice?

---
<img width="278" alt="Screenshot 2022-12-12 at 16 25 59" src="https://user-images.githubusercontent.com/842193/207147321-017631bf-de9b-416c-94cb-27be73ee248a.png">

fjorgemota replied:
Hey @alexsanford, this suggestion was suggested by @Luchadores here: p1670857311264139/1670695838.792709-slack-C8QJAM52B and here: p1670859320464819/1670695838.792709-slack-C8QJAM52B

So..yes, I think.

alexsanford commented on includes/admin/home/notices/class-sensei-home-notices.php:
```diff
@@ -54,6 +54,119 @@ public function __construct( Sensei_Home_Remote_Data_API $remote_data_api, strin
+	 * @param string $nonce The nonce to consider while checking the nonce value.
+	 * @return string An empty string, or '0', or '1'.
+	 */
+	private function get_review_answer( $nonce ) {
```
Handing in the nonce here seems a bit odd to me...is there a strong reason to do so?

To be fair, I don't think I have a strong reason _not_ to do so, except that it doesn't seem necessary (i.e. the nonce string could be given by a constant instead). In fact, given that this is a GET request and the GET parameter isn't being used to alter any data, probably we don't need a nonce here at all 🤷‍♂️

fjorgemota replied:
Hm....good question!

I passed the parameters mostly because I didn't want to create a new constant nor copy `self::HOME_NOTICE_KEY_PREFIX . 'sensei_review'` to this method again. :smile: 

Technically, you're right that nonces aren't needed, but I decided to add them mostly to avoid eventual surprises where the user visits the page and suddenly sees a happy/sad notice from us. It also helps to avoid possible issues with things like usage tracking (how many users saw the happy notice? How many users saw the sad one?), although I didn't add any on this PR.

I can remove the nonce and/or create a constant just with the notice ID, if you wish. I'm not personally strong about having a nonce, I just felt like it's a good practice given that it's changing something on the page the user is seeing.

alexsanford replied:
In this case, I don't see any serious issues with removing the nonce (the ones you outlined make sense, but I don't think they are likely to happen in practice or have high impact). But I also agree that it doesn't hurt to keep it.

If we keep it, I would personally still lean toward using a constant for the nonce action as I think it simplifies things (and is more conventional than needing to pass it around). The nonce for dismissing a notice ([in this class](https://github.com/Automattic/sensei/blob/trunk/includes/admin/class-sensei-admin-notices.php)) is a good example. Note that it doesn't need to be a dynamically created string (i.e. it doesn't need to be `$notice_id`), and probably makes more sense not to be.

fjorgemota replied:
Given your suggestion to change things in frontend instead, I decided to abandon all things related to nonce here, and so I removed this method in [5ac5933085b5089c4f382f1168c18088860b4975](https://github.com/Automattic/sensei/commit/5ac5933085b5089c4f382f1168c18088860b4975). :)

alexsanford reviewed (commented):
Works well! I left a few comments in the code.

Also, the user experience for this notice seems a bit "off" to me. It seems like the "happy path" that we want users to go through involves:

- Click "Yes".
- Wait for the page to reload.
- Click the link to add the review. Leave the review in a new tab.
- On returning to the site, the notice is still there, and we have to dismiss it.

I wonder if this could be streamlined a bit? Ideally, I think it would be something more like:

- Click "Yes".
- The "ask" for writing a review appears immediately, with no page reload needed.
- After they click the link to leave a review, the notice is automatically dismissed, and they do not see it again. (Similarly if they clicked "No" and then "Share with us how can we help")

WDYT?

Also, do we really want this notice appearing for _every_ admin on the site? Is it enough to have _one_ of the site admins leave us a review? I'm not really sure on the requirements here, so I'm just curious.

> Verify that the notice won't show for users that are not super admins;

I've confirmed that this only displays for Administrator users. Is there something specifically to test for super admins on multisite installs? Or do you just mean that this should not display for non-admins?

fjorgemota reviewed (commented):
Hey @alexsanford, thanks very much for your review!

> Also, the user experience for this notice seems a bit "off" to me. It seems like the "happy path" that we want users to go through involves:
> 
>     Click "Yes".
>     Wait for the page to reload.
>     Click the link to add the review. Leave the review in a new tab.
>     On returning to the site, the notice is still there, and we have to dismiss it.
> 
> I wonder if this could be streamlined a bit? Ideally, I think it would be something more like:
> 
>     Click "Yes".
>     The "ask" for writing a review appears immediately, with no page reload needed.
>     After they click the link to leave a review, the notice is automatically dismissed, and they do not see it again. (Similarly if they clicked "No" and then "Share with us how can we help")
> 
> WDYT?

I see what you mean, and I kinda like it. However, I personally do not see any way of implementing it without developing something like a new React component just to do that, while also looking like a traditional notice.

Yeah, we do have a Notice component on the JS side for Sensei Home already, but then we would need to adapt it so it kinda communicates with the backend to do things like:

1. Show up at the right moment; (when the condition matches)
2. Hide when the user dismissed it;
3. Dismiss it using Ajax or something.

Which...sounds way more complex than this task should be. Like, it's possible, but I don't personally like very much the idea of having this mix where a notice might be JS-only or PHP-only. It can be slightly confusing, IMHO. WDYT?

> Also, do we really want this notice appearing for every admin on the site? Is it enough to have one of the site admins leave us a review? I'm not really sure on the requirements here, so I'm just curious.

Yes! See this comment by Ronnie here: pbFul3-Ym-p2#comment-495

> I've confirmed that this only displays for Administrator users. Is there something specifically to test for super admins on multisite installs? Or do you just mean that this should not display for non-admins?

Hmmm, not necessarily. I'm worried most about this showing up for e.g. teachers, for instance, as per Ronnie's comment linked above.

alexsanford commented:
> I see what you mean, and I kinda like it. However, I personally do not see any way of implementing it without developing something like a new React component just to do that, while also looking like a traditional notice.

Maybe going "old school" and just using jQuery to hide/show versions of the notice and AJAX to submit the dismissal would work fine in this case? 🤷‍♂️

fjorgemota committed:
Fix Link to only open new tab for external links

fjorgemota committed:
Add CSS class to hide notices

fjorgemota committed:
Improve frontend logic to allow handling tasks when clicking on notice's buttons

fjorgemota committed:
Include parent_id in the normalized item returned by Sensei Home Notices Provider

fjorgemota committed:
Ignore notices with a valid parent_id on get_badge_count

fjorgemota committed:
Return all the notices on the add_review_notice and let the frontend handle it

fjorgemota committed:
Fix tests for add_review_notice given changes to the frontend logic

fjorgemota committed:
Fix Sensei Home Notices Provider tests to include parent_id

fjorgemota commented on assets/home/notices.js:
```diff
@@ -14,14 +14,54 @@ import classnames from 'classnames';
+		if ( ! action.tasks ) {
+			return;
+		}
+		for ( const task of action.tasks ) {
+			const noticeDom =
+				task.target_notice_id &&
+				document.querySelector(
+					`.sensei-notice[data-sensei-notice-id="${ task.target_notice_id }"]`
+				);
+			switch ( task.type ) {
+				case 'preventDefault':
+					e.preventDefault();
+					break;
+				case 'show':
+					noticeDom?.classList.remove( hiddenClassName );
+					break;
+				case 'hide':
+					noticeDom?.classList.add( hiddenClassName );
+					break;
+				case 'dismiss':
+					noticeDom?.querySelector( '.notice-dismiss' )?.click();
+					break;
+			}
+		}
```
The idea behind this method is to provide a way to do interactions with the other notices without having to use PHP logic to do that, making the process a little more...magical.

Right now, this is using direct DOM manipulation to change things. I thought about maybe changing React states instead, but even then we would have to simulate the click to the `notice-dismiss` link anyway, so...yeah, not sure how to handle this better.

fjorgemota replied:
Thinking a bit more about this, it might be a good idea to move this "event handler" to the same script where we handle the click event for the ".notice-dismiss" link, and passing the tasks list via a data attribute instead.

This way, it would be possible to make this logic work even with the old native Sensei notices HTML, which is.. Better than having this exclusive to the notices on Sensei home, at least, while also removing this non react logic from a react component, too.

WDYT?

fjorgemota reviewed (commented):
Hey @alexsanford, thanks for your suggestions.

I changed the PR to implement kinda what you proposed. It now shows/hides the notices appropriately on the frontend, but I relied on some additional code to avoid having to manipulate each notice directly.

I await your feedback on what you think of this solution. If you don't like it and/or prefer a more straightforward solution, let me know. 

Thanks.

alexsanford commented on assets/home/notices.js:
```diff
@@ -14,14 +14,54 @@ import classnames from 'classnames';
+ *
+ * @param {Object} action       The action to return the event listener for.
+ * @param {Array}  action.tasks The tasks to run.
+ * @return {Function} The event listener that runs the specified tas.s
```
I think this line has a typo.

fjorgemota replied:
Good catch! I moved the logic out of this method in [c1cbcc6](https://github.com/Automattic/sensei/pull/6274/commits/c1cbcc67ce75db84bf6313b9263ad9e124de6983) - So this doc wasn't necessary anymore. Thanks for the catch, though!

alexsanford commented on includes/admin/home/notices/class-sensei-home-notices-provider.php:
```diff
@@ -81,10 +81,13 @@ public function get( $max_age = HOUR_IN_SECONDS ): array {
-		$notices = $this->get( DAY_IN_SECONDS );
+		$notices                 = $this->get( DAY_IN_SECONDS );
+		$notices_with_parent_ids = array_column( $notices, 'parent_id' );
+		$notices_with_parent_ids = array_filter( $notices_with_parent_ids );
```
Maybe it would be clearer to just do `$notices_with_parent_ids = array_filter( array_column( $notices, 'parent_id' ) );` rather than reassigning the variable? 🤷‍♂️

fjorgemota replied:
Sure, done in [1701b7e](https://github.com/Automattic/sensei/pull/6274/commits/1701b7e2f6f6a1a98c54ed748e0e514c0e315bf1). Thanks!

alexsanford reviewed (dismissed):
The UX is much nicer! 🙂 

I'm not sure how I feel about the complexity of the implementation, tbh. Seems like it would be warranted if we knew that we were going to be using this type of interaction a lot. Although if that was the case, maybe just going full-React would be better at that point.

But given that this is a pretty simple feature, I don't feel the need to be super picky about the implementation. It seems sound, and works well, so let's 🚢 it! 🙂

fjorgemota committed:
Add support for passing data attributes to Link components

fjorgemota committed:
Only set UTM tags if they weren't set yet

fjorgemota committed:
Move logic to handle tasks on buttons in notices to sensei-notice-dismiss.js

fjorgemota committed:
Add data-sensei-notice-tasks support to class-sensei-admin-notices.php

fjorgemota committed:
Use notice_id instead of target_notice_id as parameter name

fjorgemota committed:
Make get_badge_count method a little bit more clear based on Alex's suggestion

Co-authored-by: alexsanford <alexsanford@users.noreply.github.com>

fjorgemota committed:
Remove code that overwrote show_after on Sensei_Home_Notices

fjorgemota commented:
Hey @alexsanford, thanks for the review.

Yeah, I kinda feel a little conflicted about the complexity of the implementation, as I just didn't want to add custom code to Sensei JUST to handle this special...notice. It sounds like a weird scenario to have, and I personally don't like the increased mental load that comes with making code specific based on things like IDs/classes appearing out of nowhere in the code. For instance, I looked for a good few minutes where/how the dismiss button was appearing until I discovered that it was handled in a special way by WordPress: https://developer.wordpress.org/reference/hooks/admin_notices/ 

I improved the implementation by moving the logic to the same file where we handle the click-on notice-dismiss, so we're avoiding changes to React components while making it a more useful outside Sensei Home too. This is done in [c1cbcc6](https://github.com/Automattic/sensei/pull/6274/commits/c1cbcc67ce75db84bf6313b9263ad9e124de6983), [8df3a46](https://github.com/Automattic/sensei/pull/6274/commits/8df3a4632c40a48084c5d4868f037be74230678d), [62e748b](https://github.com/Automattic/sensei/pull/6274/commits/62e748bf3949373d049a3b56d037526e833cdbf3) and  [46aace9](https://github.com/Automattic/sensei/pull/6274/commits/46aace92de6f0ca9ea5d53fbf01dee5b1f3a8937) (this last one was just to avoid the looong name `target_notice_id`).

I also found an issue where the UTM tags were being overwritten, and fixed it in [bcd9c9b](https://github.com/Automattic/sensei/pull/6274/commits/bcd9c9b45aca91c3d11298e2c610e5c9a2f2ac34). 

Finally, I also implemented your suggestion about `get_badge_count` in [1701b7e](https://github.com/Automattic/sensei/pull/6274/commits/1701b7e2f6f6a1a98c54ed748e0e514c0e315bf1).

Sorry for the long(er) PR. These are my last changes to this PR unless you find something else. And thanks again for your review and patience. :|

alexsanford reviewed (approved):
Re-tested, works well!
